File: app/src/main/java/pl/futuredev/capstoneproject/ui/adapters/TopPlacesAdapter.java
package pl.futuredev.capstoneproject.ui.adapters;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.squareup.picasso.Picasso;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import pl.futuredev.capstoneproject.R;
import pl.futuredev.capstoneproject.models.Image;
import pl.futuredev.capstoneproject.models.Medium;
import pl.futuredev.capstoneproject.models.Original;
import pl.futuredev.capstoneproject.models.Result;
import pl.futuredev.capstoneproject.models.Thumbnail;
import pl.futuredev.capstoneproject.ui.IOnClickHandler;

public class TopPlacesAdapter extends RecyclerView.Adapter<TopPlacesAdapter.ViewHolder> {


    private List<Result> resultList;
    private IOnClickHandler iOnClickHandler;

    public TopPlacesAdapter(List<Result> resultList, IOnClickHandler iOnClickHandler) {
        this.resultList = resultList;
        this.iOnClickHandler = iOnClickHandler;
    }

    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        @BindView(R.id.iv_image)
        ImageView ivImage;
        @BindView(R.id.tv_name)
        TextView tvName;
        @BindView(R.id.tv_snippet)
        TextView tvSnippet;

        public ViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {
            int clickPosition = getAdapterPosition();
            iOnClickHandler.onClick(clickPosition);
        }
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.top_places_single_layout, parent, false);
        view.setFocusable(true);
        return new ViewHolder(view);
    }


    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {

        ImageView ivImage = holder.ivImage;
        TextView tvName = holder.tvName;
        TextView tvSnippet = holder.tvSnippet;

        List<Image> images = resultList.get(position).getImages();
        if (images != null && !images.isEmpty()) {
            Medium mediumImage = images.get(0).getSizes().getMedium();

            Picasso.get()
                    .load(mediumImage.getUrl())
                    .into(ivImage, new com.squareup.picasso.Callback() {
                        @Override
                        public void onSuccess() {
                            Picasso.get().load(mediumImage.getUrl()).into(ivImage);

                        }

                        @Override
                        public void onError(Exception e) {
                            Picasso.get().load(R.drawable.rest1).into(ivImage);
                        }
        });

        }

        tvName.setText(resultList.get(position).getName());
        tvSnippet.setText(resultList.get(position).getSnippet());
    }

    @Override
    public int getItemCount() {
        return resultList.size();
    }
}

File: app/src/main/java/pl/futuredev/capstoneproject/ui/adapters/TopPlacesAdapter.kt
package pl.futuredev.capstoneproject.ui.adapters

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.RequestManager
import pl.futuredev.capstoneproject.R
import pl.futuredev.capstoneproject.data.remote.entities.Image
import pl.futuredev.capstoneproject.data.remote.entities.Result
import pl.futuredev.capstoneproject.databinding.AdapterTopPlacesBinding
import javax.inject.Inject

class TopPlacesAdapter @Inject constructor(
    private val glide: RequestManager,
    private var cities: List<Result>
) :
    RecyclerView.Adapter<TopPlacesAdapter.TopPlacesViewHolder>() {

    private var onItemClickListener: ((Int) -> Unit)? = null

    class TopPlacesViewHolder(private val itemBinding: AdapterTopPlacesBinding) :
        RecyclerView.ViewHolder(itemBinding.root) {

        val name = itemBinding.tvName
        val snippet = itemBinding.tvSnippet
        val image = itemBinding.ivImage
        val details = itemBinding.ivDetails
        val score = itemBinding.tvScore
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TopPlacesViewHolder {
        val itemBinding =
            AdapterTopPlacesBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return TopPlacesViewHolder(itemBinding)

    }

    override fun onBindViewHolder(holder: TopPlacesViewHolder, position: Int) {
        val city = cities[position]
        val images: List<Image> = cities[position].images

        holder.itemView.apply {
            holder.name.text = city.name
            holder.snippet.text = city.snippet
            if (images.isNotEmpty()) {
                glide.load(city.images[0].sizes.original.url).into(holder.image)
                if (images[0].ownerUrl.isEmpty()) {
                    holder.details.visibility = View.GONE
                } else {
                    glide.load(R.drawable.more_details).into(holder.details)
                }
            }

            setOnClickListener {
                onItemClickListener?.let { click ->
                    click(position)
                }
            }
            val scores = cities[position].score.toInt().toString()
            holder.score.text = "Scores: $scores/10"
        }
    }

    override fun getItemCount(): Int {
        return cities.size
    }

    fun setOnItemClickListener(onItemClickListener: (Int) -> Unit) {
        this.onItemClickListener = onItemClickListener
    }

}